pkg/apis/application/v1alpha1/source_integrity.go

```diff
@@ -0,0 +1,251 @@
+package v1alpha1
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"slices"
+	"strings"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/argoproj/argo-cd/v3/util/git"
+)
+
+type SourceIntegrity struct {
+	// Git - policies for git source verification
+	// Mandatory field until there are alternatives
+	Git *SourceIntegrityGit `json:"git" protobuf:"bytes,1,name=git"`
+}
+
+type SourceIntegrityGitFunc func(gitClient git.Client, unresolvedRevision string) (*SourceIntegrityCheckResult, error)
+
+func (f SourceIntegrityGitFunc) Verify(gitClient git.Client, unresolvedRevision string) (*SourceIntegrityCheckResult, error) {
+	if f == nil {
+		return nil, nil
+	}
+	return f(gitClient, unresolvedRevision)
+}
+
+// ForGit evaluate if there are cheks for the git sources per given ApplicationSource and returns a function performing such check.
+// If there are no cheks for git or the application source, this returns nil.
+// This indirection is needed to detect the existence of relevant criteria without their actual execution.
+func (si *SourceIntegrity) ForGit(repoURL string) SourceIntegrityGitFunc {
+	if si == nil || si.Git == nil {
+		return nil
+	}
```

Comment:
Instead of `ForGit`, could we not make this more general? Something like 


```suggestion
// SourceIntegrityFunc renamed to reflect that it's more general
func (si *SourceIntegrity) FromRepo(*v1alpha1.Repository repo) SourceIntegrityFunc {
	if repo == nil || repo.Type != "git" |  si == nil || si.Git == nil {
		return nil
	}
	//existing logic
```

In that case we'd also need something equivalent for `ApplicationSource` which would do pretty much the same thing

```suggestion
// SourceIntegrityFunc renamed to reflect that it's more general
func (si *SourceIntegrity) FromSource(*v1alpha1.ApplicationSource source) SourceIntegrityFunc {
    // For now it's Git only (alternatively add a simple check to see if the repo url is a git one)
	if source == nil || si == nil || si.Git == nil {
		return nil
	}
	//existing logic as FromRepo
```

Reply:
I have evolved the API differently in the end. There were several considerations for this:

- The api now accepts multiple sources only when it detects if there are some criteria (`func HasCriteria(si *v1alpha1.SourceIntegrity, sources ...v1alpha1.ApplicationSource) bool`).
- The actual verification always
  - Operates with a single source, so there is no reason to accept multiple sources there.
  - Requires specifics like git client with particular repo, so the call has to be both specific to git and the repo: `func VerifyGit(si *v1alpha1.SourceIntegrity, gitClient git.Client, unresolvedRevision string) (result *v1alpha1.SourceIntegrityCheckResult, err error)`. We will add sibling functions `VerifyXYZ` for the future source integrity targets, the parameters will be specific to whatever it will do.
- There is no need to expose the function anymore - looked a bit ugly.